themes/default/content/blog/iac-recommended-practices-developer-stacks-git-branches/index.md

```diff
@@ -0,0 +1,209 @@
+
+## Stepping back to reassess
+
+With what we've learned so far, we can return to the questions we raised at the beginning of this post and begin to sketch out a few recommendations for Team Zephyr. Those questions, again, were:
+
+* **How can I best enable multiple developers to collaborate on a Pulumi project?** For Zephyr, an ideal answer to this question would give every member of the team the ability to iterate on both application and infrastructure code concurrently, deploying as often as needed without having to worry about disrupting the work or the velocity of anyone else on the team.
+
+* **How can I use Git and Git branching to support this collaboration?** So far, the feature-branch/pull-request workflow has been serving the Zephyr team well. Should they continue on the same path, merging frequently to `main` and deploying to `dev` and `prod` from there? Or is there some alternative workflow that might serve the team better, either now or at some time in the future?
+
+Let's answer these questions now by taking a look at how the Zephyr team handled them.
+
+### Using developer stacks
+
+We've seen a few ways that having multiple developers share a single stack for ad-hoc deployment can lead to various problems. But we've also seen at least one situation in which being able to run ad-hoc deployments --- without the risk of colliding with existing environments --- can be quite valuable. In fact there are many reasons you might want to do this --- and fortunately Pulumi makes it quite easy to do so.
+
+The recommended practice is to allow developers to create and manage their own Pulumi stacks. In this workflow, developers working on feature branches are able to iterate freely on both application and infrastructure, deploying as often as necessary into a shared cloud environment (typically a sandbox environment, isolated from production) and destroying cloud infrastructure when it's no longer needed.
+
+For the developer, the typical workflow looks something like this:
+
+1. Check out a new feature branch
+2. Make changes to application and/or infrastructure code
+3. Select your personal developer stack with `pulumi stack select`
+4. Deploy your stack into your team's shared cloud environment with `pulumi up`
+5. Commit your code and submit a pull request, optionally including a link either to the Pulumi update (e.g., when using the Pulumi Service) or to certain key resources belonging to the stack --- for example, an application's primary HTTP endpoint
+
+The benefits of a workflow like this one are many:
+
+* No more resource-naming collisions. Because Pulumi names cloud resources [uniquely by default](/docs/intro/concepts/resources/names/#autonaming), no two stacks will ever produce cloud resources with the same names.
+* No more deployment conflicts. When every developer has a stack of their own, they can deploy to the sandboxed environment as often as necessary, without having to wait for another deployment to finish first.
```

Comment:
also when you are talking about stacks here you are talking about stacks within a project correct? we aren't talking about creating a completely separate project right?

Reply:
Correct -- one project, multiple stacks. Later in the series we'll break out another project, but for now, it's just the one.